lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs

```diff
@@ -2203,25 +2198,29 @@ instance Arbitrary (MixedSign Value) where
         genPositive = arbitrary
     shrink (MixedSign v) = MixedSign <$> shrink v
 
-instance forall era. IsRecentEra era => Arbitrary (PartialTx era) where
-    arbitrary = do
-        tx <- genTxForBalancing
-        extraUTxO <- genExtraUTxO (txInputs tx)
-        let redeemers = []
-        let timelockKeyWitnessCounts = mempty
-        pure PartialTx {tx, extraUTxO, redeemers, timelockKeyWitnessCounts}
-      where
-        genExtraUTxO :: Set TxIn -> Gen (UTxO era)
-        genExtraUTxO = fmap UTxO . genMapFromKeysWith genTxOut
-        txInputs :: Tx era -> Set TxIn
-        txInputs tx = tx ^. bodyTxL . inputsTxBodyL
-    shrink partialTx@PartialTx {tx, extraUTxO} =
-        [ partialTx {extraUTxO = extraUTxO'}
-        | extraUTxO' <- shrinkInputResolution extraUTxO
-        ] <>
-        [ restrictResolution (partialTx {tx = tx'})
-        | tx' <- shrinkTx tx
-        ]
+instance IsRecentEra era => Arbitrary (PartialTx era) where
+    arbitrary = mkPartialTx <$> genTxWithUTxO
+    shrink = shrinkMapBy mkPartialTx unPartialTx shrinkTxWithUTxO
+
+mkPartialTx :: IsRecentEra era => TxWithUTxO era -> PartialTx era
+mkPartialTx (TxWithUTxO tx extraUTxO) =
+    PartialTx {tx, extraUTxO, redeemers = [], timelockKeyWitnessCounts = mempty}
+
+unPartialTx :: IsRecentEra era => PartialTx era -> TxWithUTxO era
+unPartialTx PartialTx {tx, extraUTxO} =
+    TxWithUTxO.constructFiltered tx extraUTxO
+
+genTxWithUTxO :: IsRecentEra era => Gen (TxWithUTxO era)
+genTxWithUTxO = TxWithUTxO.generate genTxForBalancing genTxIn genTxOut
+  where
+    genTxIn :: Gen TxIn
+    genTxIn = fromWalletTxIn <$> W.genTxIn
+
+shrinkTxWithUTxO :: IsRecentEra era => TxWithUTxO era -> [TxWithUTxO era]
+shrinkTxWithUTxO = TxWithUTxO.shrinkWith shrinkTx shrinkUTxOToSubsets
+  where
+    shrinkUTxOToSubsets :: IsRecentEra era => UTxO era -> [UTxO era]
+    shrinkUTxOToSubsets = shrinkMapBy UTxO unUTxO shrinkMapToSubmaps
```

Comment:
Could you move these to `where` clauses of the `Arbitrary (PartialTx era)`? I think it would be cleaner and more clear where they're used.

Reply:
Hi @Anviking 

It's not possible to move `mkPartialTx`, because it is used by both `arbitrary` and `shrink`.

AFAIK it's not possible to have a `where` clause that is common to the entire instance definition.

Reply:
Though I have renamed these functions in: 5262ce259cb9db4c4d6366be0bcaafa134138b0c.